internal/controller/postgrescluster/cluster.go

```diff
@@ -189,6 +190,59 @@ func (r *Reconciler) reconcileClusterPrimaryService(
 	return err
 }
 
+// generateClusterReplicaServiceIntent returns a v1.Service that exposes
+// PostgreSQL replica instances.
+func generateClusterReplicaServiceIntent(cluster *v1beta1.PostgresCluster) *corev1.Service {
+	service := &corev1.Service{ObjectMeta: naming.ClusterReplicaService(cluster)}
+	service.SetGroupVersionKind(corev1.SchemeGroupVersion.WithKind("Service"))
+
+	service.Annotations = naming.Merge(
+		cluster.Spec.Metadata.GetAnnotationsOrNil())
+	service.Labels = naming.Merge(
+		cluster.Spec.Metadata.GetLabelsOrNil(),
+		map[string]string{
+			naming.LabelCluster: cluster.Name,
+			naming.LabelRole:    naming.RoleReplica,
+		})
+
+	// Allocate an IP address and let Kubernetes manage the Endpoints by selecting
+	// Pods with the replica role.
+	// - https://docs.k8s.io/concepts/services-networking/service/#defining-a-service
+	service.Spec.Type = corev1.ServiceTypeClusterIP
+	service.Spec.Selector = map[string]string{
+		naming.LabelCluster: cluster.Name,
+		naming.LabelRole:    naming.RoleReplica,
+	}
+
+	// The TargetPort must be the name (not the number) of the PostgreSQL
+	// ContainerPort. This name allows the port number to differ between Pods,
+	// which can happen during a rolling update.
+	service.Spec.Ports = []corev1.ServicePort{{
+		Name:       naming.PortPostgreSQL,
+		Port:       *cluster.Spec.Port,
+		Protocol:   corev1.ProtocolTCP,
+		TargetPort: intstr.FromString(naming.PortPostgreSQL),
+	}}
+
+	return service
+}
+
+// +kubebuilder:rbac:groups="",resources="services",verbs={create,patch}
+
+// reconcileClusterReplicaService writes the Service that exposes PostgreSQL
+// replica instances.
+func (r *Reconciler) reconcileClusterReplicaService(
+	ctx context.Context, cluster *v1beta1.PostgresCluster,
+) error {
+	service := generateClusterReplicaServiceIntent(cluster)
+
+	err := errors.WithStack(r.setControllerReference(cluster, service))
```

Comment:
We have been setting the controller reference inside of generate intent functions. Is there an argument to do it this way, i.e. generate the intent using one function, then set the controller reference outside of that function :question:

Reply:
With it outside of the generate function, the generate function returns no errors and does not depend on Reconciler.

Reply:
Good thought/question, I too would consider the controller reference as part of the intent.

Reply:
The `SetControllerReference` helper provided by `controller-runtime` is geared toward the create-or-update pattern there. It returns errors when there's already a controller attached to the object (the Go struct) or you try to set ownership across namespaces.

We're doing SSA, so the Go struct always lacks a controller. That validation will never fail, and the upsert it performs is also always an append. 😐 

https://github.com/kubernetes-sigs/controller-runtime/blob/v0.8.3/pkg/controller/controllerutil/controllerutil.go#L59

🌱 At some point, we'll want (need?) to be using [applyconfigurations](https://pkg.go.dev/k8s.io/client-go/applyconfigurations) to build our SSA intents.

Reply:
To me it simply feels like a "generate intent" function should generate the full intent, not just part of it - otherwise the function feels confusing.  I know we've also done similar with other/similar functions (i.e. moved setting the controller ref into the "generate intent" function).

If that helper function is the reason we can't or don't want to, maybe we should ditch it?

Reply:
And yes ... I want applyconfigurations!

Any updates out there as to when we'll see something similar in controller runtime?

Reply:
> Any updates out there as to when we'll see something similar in controller runtime?

Not that I can see.

Reply:
> We have been setting the controller reference inside of generate intent functions.

> To me it simply feels like a "generate intent" function should generate the full intent, not just part of it - otherwise the function feels confusing.

Moved in [`ee784a1c`](https://github.com/CrunchyData/postgres-operator/pull/2596/commits/ee784a1c3f8940891722123ac169df78a777e61d).